core/arch/arm/kernel/boot.c

```diff
@@ -1256,14 +1259,55 @@ void __weak boot_init_primary_early(unsigned long pageable_part,
 				    unsigned long nsec_entry __maybe_unused)
 {
 	unsigned long e = PADDR_INVALID;
+	struct transfer_list_entry *te = NULL;
+	struct transfer_list_header *tl = NULL;
+	struct dt_descriptor *dt = get_external_dt_desc();
 
 #if !defined(CFG_WITH_ARM_TRUSTED_FW)
 	e = nsec_entry;
 #endif
+	/* When a TL is passed in, pageable_part has been forced to 0 */
+	if (IS_ENABLED(CFG_TRANSFER_LIST) && transfer_list_pa &&
+	    !pageable_part) {
+		/* map the TL */
+		tl = transfer_list_map(transfer_list_pa);
+		if (!tl)
+			panic("Failed to map transfer list");
+		dt->tl = tl; /* save the mapped TL */
+
+		te = transfer_list_find(tl, TL_TAG_OPTEE_PAGABLE_PART);
+		if (IS_ENABLED(CFG_WITH_PAGER) && te)
+			pageable_part = get_le64(transfer_list_entry_data(te));
+	}
 
 	init_primary(pageable_part, e);
 }
 
+#if defined(CFG_TRANSFER_LIST)
+/*
+ * Note: this function is weak just to make it possible to exclude it from
+ * the unpaged area so that it lies in the init area.
+ */
+void __weak boot_save_transfer_list(unsigned long transfer_list,
+				    unsigned long fdt)
+{
+	struct transfer_list_header *tl = (void *)transfer_list;
+	struct transfer_list_entry *te = NULL;
+
+	if (!IS_ALIGNED(transfer_list, TL_ALIGNMENT_FROM_ORDER(tl->alignment)))
```

Comment:
sorry I missed this one. This line checks tl alignment. We can't get `tl->aligment` if `tl` is not aligned.
I think this needs a 2 step test:
```c
	if (!IS_ALIGNED_WITH(transfer_list, struct transfer_list_header) ||
	    !IS_ALIGNED(transfer_list, TL_ALIGNMENT_FROM_ORDER(tl->alignment)))
		panic("Transfer list base address is not aligned");

Reply:
Yes, you are right, I missed this.
Fixed.